camera/camera_balsaq.cc

```diff
@@ -8,32 +8,58 @@
 #include <iostream>
 #include <sstream>
 
-//%deps(balsa_queue)
-//%deps(message)
-
 namespace jet {
 
-constexpr double WEBCAM_EXPOSURE = 0.01;
+namespace {
+
+// Generate a config structure from a yaml config
+//
+// Asserts that the required config fields are actually set.
+// Has no defaulting behavior.
+CameraConfiguration generate_capture_config(const Config& config) {
+  assert(config["frames_per_second"]);
+  assert(config["exposure"]);
+  assert(config["auto_focus"]);
+  assert(config["auto_exposure"]);
+  assert(config["width_pixels"]);
+  assert(config["height_pixels"]);
+
+  CameraConfiguration camera_config;
+  {
+    camera_config.frames_per_second = config["frames_per_second"].as<int>();
+    camera_config.exposure = config["exposure"].as<double>();
+    camera_config.auto_focus = config["auto_focus"].as<int>();
+    camera_config.auto_exposure = config["auto_exposure"].as<double>();
+    camera_config.width_pixels = config["width_pixels"].as<int>();
+    camera_config.height_pixels = config["height_pixels"].as<int>();
+  }
+  return camera_config;
+}
+}  // namespace
 
 void CameraBq::init(const Config& config) {
-  Camera camera;
-  std::cout << "Camera BQ: camera serial " << config["serial_number"].as<std::string>() << std::endl;
-  camera = camera_manager_.get_camera(config["serial_number"].as<std::string>());
+  const Camera camera = camera_manager_.get_camera(config["serial_number"].as<std::string>());
   camera_serial_number_ = camera.serial_number;
+  std::cout << "Camera BQ: camera serial " << config["serial_number"].as<std::string>() << std::endl;
   cap = cv::VideoCapture(camera.video_index);
-  cap.set(cv::CAP_PROP_FRAME_WIDTH, 640);
-  cap.set(cv::CAP_PROP_FRAME_HEIGHT, 480);
-  cap.set(cv::CAP_PROP_FPS, CAMERA_FPS);
+
+  camera_config_ = generate_capture_config(config);
+
+  cap.set(cv::CAP_PROP_FRAME_WIDTH, camera_config_.width_pixels);
+  cap.set(cv::CAP_PROP_FRAME_HEIGHT, camera_config_.height_pixels);
+  cap.set(cv::CAP_PROP_FPS, camera_config_.frames_per_second);
+
+  // 0 is the id of video device.0 if you have only one camera.
```

Comment:
Not immediately clear what this refers to

Reply:
Not sure where this came from